core/cap-0063.md

```diff
@@ -0,0 +1,447 @@
+```
+CAP: 0063
+Title: Soroban In-memory Read Resource
+Working Group:
+    Owner: Garand Tyson <@SirTyson>
+    Authors: Garand Tyson <@SirTyson>
+    Consulted: Dmytro Kozhevin <@dmkozh>, Nicolas Barry <@MonsieurNicolas>
+Status: Draft
+Created: 2024-12-09
+Discussion: https://github.com/stellar/stellar-protocol/discussions/1585
+Protocol version: 23
+```
+
+## Simple Summary
+
+This proposal introduces a new resource type for Soroban reads, distinguishing between in-memory and
+disk reads. This also proposes automatic restoration for archived entries via `InvokeHostFunctionOp`.
+
+## Working Group
+
+As specified in the Preamble.
+
+## Motivation
+
+By distinguishing disk and in-memory reads, this proposal allows for significant increase in Soroban
+read limits. This distinction also allows for safe automatic entry restoration, significantly improving UX.
+
+### Goals Alignment
+
+This change is aligned with the goal of lowering the cost and increasing the scale of the network.
+
+## Abstract
+
+[CAP-0062](cap-0062.md) introduces partial State Archival, where evicted and live Soroban state is stored in separate databases.
+This separation allows live Soroban state to be efficiently cached entirely in memory, removing disc reads entirely for all
+live Soroban state. However, classic state and evicted entries are still subject to disk reads.
+
+This CAP introduces a new resource type for reads, distinguishing between disk reads and in-memory reads. By making this
+distinction at the protocol level, read limits for Soroban data can greatly increase. This also opens the door for other
+optimizations, such as a complete module cache for all live contracts.
+
+Additionally, this CAP introduces automatic restoration via `InvokeHostFunctionOp`, where any archived key present in the
+footprint is automatically restored. Initially in protocol 20, the state archival design was not solidified enough to
+enable automatic restoration, so an explicit restore operation was required. While this was not technically required in
+protocol 20, it was important for contract interfaces to reason properly about restoration such that when full state
+archival was introduce, it would not break preexisting deployments. Given that the full State Archival design has been
+mostly finalized in [CAP-0057](cap-0057.md), it is now appropriate to introduce automatic restoration.
+
+## Specification
+
+### XDR changes
+
+```
+--- a/Stellar-contract-config-setting.x
++++ b/Stellar-contract-config-setting.x
+@@ -60,6 +60,46 @@ struct ConfigSettingContractLedgerCostV0
+     uint32 bucketListWriteFeeGrowthFactor;
+ };
+ 
++// Ledger access settings for contracts.
++struct ConfigSettingContractLedgerCostV1
```

Comment:
Is it possible to split limits and fees in different structs? That way the one with fees can have cost in the name (ConfigSettingContractLedgerCost) while the other one could be named ConfigSettingContractLedgerLimitV1.

(Probably a question for @dmkozh)

Reply:
Originally I did that, but Dima suggested one struct instead (I unresolved the comment above for context). I don't care particularly strongly either way.

Reply:
> Is it possible to split limits and fees in different structs?

We didn't do that initially, and I don't think it makes much sense to change this now. Conceptually the 'cost' here concerns both resource limits ('cost' for the validators) and fees ('cost' for the user).

> Originally I did that, but Dima suggested one struct instead 

FWIW I suggested to use one struct for the setting version migration, this wasn't about separating the fees and limits. But this is the first time we do a setting update and there is no well-defined process yet. Giving this a second thought, I think going the extension way is probably less disruptive for anyone who consumes these settings, as having two versions introduces the risk of using the deprecated value. The name should be `ConfigSettingContractLedgerCostV0Ext` then. Version bumps should probably be reserved for non-incremental and truly breaking changes. Sorry for the confusion, I should have thought more about this initially.